geopandas/tools/overlay.py

```diff
@@ -331,69 +340,69 @@ def _make_valid(df):
-            )
+    result.reset_index(drop=True, inplace=True)
+    return result
 
-            # level_0 created with above reset_index operation
-            # and represents the original geometry collections
-            # TODO avoiding dissolve to call union_all in this case could further
-            # improve performance (we only need to collect geometries in their
-            # respective Multi version)
-            dissolved = exploded.dissolve(by="level_0")
-            result.loc[is_collection, geom_col] = dissolved[geom_col].values
-        else:
-            num_dropped_collection = 0
-
-        # Now we filter all geometries (in theory we don't need to do this
-        # again for the rows handled above for GeometryCollections, but filtering
-        # them out is probably more expensive as simply including them when this
-        # is typically about only a few rows)
-        orig_num_geoms = result.shape[0]
-        if geom_type in polys:
-            result = result.loc[result.geom_type.isin(polys)]
-        elif geom_type in lines:
-            result = result.loc[result.geom_type.isin(lines)]
-        elif geom_type in points:
-            result = result.loc[result.geom_type.isin(points)]
-        else:
-            raise TypeError("`keep_geom_type` does not support {}.".format(geom_type))
-        num_dropped = orig_num_geoms - result.shape[0]
-
-        if (num_dropped > 0 or num_dropped_collection > 0) and keep_geom_type_warning:
-            warnings.warn(
-                "`keep_geom_type=True` in overlay resulted in {} dropped "
-                "geometries of different geometry types than df1 has. "
-                "Set `keep_geom_type=False` to retain all "
-                "geometries".format(num_dropped + num_dropped_collection),
-                UserWarning,
-                stacklevel=2,
-            )
 
-    result.reset_index(drop=True, inplace=True)
+def _collection_extract(df, geom_type, keep_geom_type_warning):
+    # Check input
+    if geom_type in ["Polygon", "MultiPolygon"]:
+        geom_types = ["Polygon", "MultiPolygon"]
+    elif geom_type in ["LineString", "MultiLineString", "LinearRing"]:
+        geom_types = ["LineString", "MultiLineString", "LinearRing"]
+    elif geom_type in ["Point", "MultiPoint"]:
+        geom_types = ["Point", "MultiPoint"]
+    else:
+        raise TypeError(f"`geom_type` does not support {geom_type}.")
+
+    result = df.copy()
+
+    # First we filter the geometry types inside GeometryCollections objects
+    # (e.g. GeometryCollection([polygon, point]) -> polygon)
+    # we do this separately on only the relevant rows, as this is an expensive
+    # operation (an expensive no-op for geometry types other than collections)
+    is_collection = result.geom_type == "GeometryCollection"
+    if is_collection.any():
+        geom_col = result._geometry_column_name
+        collections = result[[geom_col]][is_collection]
```

Comment:
```suggestion
        collections = result.loc[is_collection, [geom_col]]
```
I know you didn't write this code it just shows up again in the diff, but may as well be consistent on using loc and avoiding chained assignment.